embedding/cpp/flutter_engine.cc

```diff
@@ -0,0 +1,134 @@
+// Copyright 2022 Samsung Electronics Co., Ltd. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "include/flutter_engine.h"
+
+#include "tizen_log.h"
+#include "utils.h"
+
+std::unique_ptr<FlutterEngine> FlutterEngine::Create(
+    const std::optional<std::string>& dart_entrypoint,
+    const std::optional<std::vector<std::string>>& dart_entrypoint_args) {
+  return std::move(FlutterEngine::Create(
+      "../res/flutter_assets", "../res/icudtl.dat", "../lib/libapp.so",
+      dart_entrypoint, dart_entrypoint_args));
+}
+
+std::unique_ptr<FlutterEngine> FlutterEngine::Create(
+    const std::string& assets_path, const std::string& icu_data_path,
+    const std::string& aot_library_path,
+    const std::optional<std::string>& dart_entrypoint,
+    const std::optional<std::vector<std::string>>& dart_entrypoint_args) {
+  FlutterEngine* engine =
+      new FlutterEngine(assets_path, icu_data_path, aot_library_path,
+                        dart_entrypoint, dart_entrypoint_args);
+  if (engine->engine_) {
+    return std::unique_ptr<FlutterEngine>(engine);
+  } else {
+    delete engine;
+    return nullptr;
+  }
+}
+
+FlutterEngine::FlutterEngine(
+    const std::string& assets_path, const std::string& icu_data_path,
+    const std::string& aot_library_path,
+    const std::optional<std::string>& dart_entrypoint,
+    const std::optional<std::vector<std::string>>& dart_entrypoint_args) {
+  FlutterDesktopEngineProperties engine_prop = {};
+  engine_prop.assets_path = assets_path.c_str();
+  engine_prop.icu_data_path = icu_data_path.c_str();
+  engine_prop.aot_library_path = aot_library_path.c_str();
+
+  // Read engine arguments passed from the tool.
+  Utils::ParseEngineArgs(&engine_arguments_);
+  std::vector<const char*> switches;
+  for (auto& arg : engine_arguments_) {
+    switches.push_back(arg.c_str());
+  }
+  engine_prop.switches = switches.data();
+  engine_prop.switches_count = switches.size();
+
+  engine_prop.entrypoint =
+      dart_entrypoint.has_value() ? dart_entrypoint.value().c_str() : nullptr;
+
+  std::vector<const char*> entrypoint_args;
+  if (dart_entrypoint_args.has_value()) {
+    for (const std::string& arg : dart_entrypoint_args.value()) {
+      entrypoint_args.push_back(arg.c_str());
+    }
+  }
+  engine_prop.dart_entrypoint_argc = entrypoint_args.size();
+  engine_prop.dart_entrypoint_argv = entrypoint_args.data();
+
+  engine_ = FlutterDesktopEngineCreate(engine_prop);
+}
+
+FlutterEngine::~FlutterEngine() {
+  if (owns_engine_) {
+    Shutdown();
+  }
+}
+
+bool FlutterEngine::Run() {
+  if (engine_) {
+    is_running_ = FlutterDesktopEngineRun(engine_);
+  }
+  return is_running_;
+}
+
+void FlutterEngine::Shutdown() {
+  if (engine_) {
+    FlutterDesktopEngineShutdown(engine_);
```

Comment:
Maybe we should set `is_running_` to false (or even set `engine_` to nullptr) after terminating the engine (although it might not be very necessary).

Reply:
done.

Reply:
Thanks. Setting `engine_` to nullptr also affects other methods in this class so please be careful.

Reply:
Thank you. I've added more guards.